lib/auth/auth.go

```diff
@@ -2511,9 +2520,49 @@ func (a *Server) mfaAuthChallenge(ctx context.Context, user string, u2fStorage u
 			})
 		}
 	}
+	if len(groupedDevs.Webauthn) > 0 {
+		webLogin := &wanlib.LoginFlow{
+			U2F:      u2fPref,
+			Webauthn: webConfig,
+			Identity: a.Identity,
+		}
+		assertion, err := webLogin.Begin(ctx, user)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		challenge.WebauthnChallenge = wanlib.CredentialAssertionToProto(assertion)
+	}
+
 	return challenge, nil
 }
 
+type devicesByType struct {
+	TOTP     bool
+	U2F      []*types.MFADevice
+	Webauthn []*types.MFADevice
+}
+
+func groupByDeviceType(devs []*types.MFADevice, groupU2F, groupWebauthn bool) devicesByType {
+	res := devicesByType{}
+	for _, dev := range devs {
+		switch dev.Device.(type) {
+		case *types.MFADevice_Totp:
+			res.TOTP = true
```

Comment:
Why do we only have a single bool flag for TOTP devices?

Reply:
This is the legacy behavior - TOTP is either enabled or not.